### PR TITLE
Checkout: Add Payment Method iDEAL

### PR DESCRIPTION
**Pull request by akirk**
Adding support for iDEAL in the Netherlands.

Depends on #19006 (which was extracted from this PR)

payment-methods

### CODE REVIEW

File: client/my-sites/checkout/checkout/credit-card-payment-box.jsx [hunk 1]
```diff
@@ -114,8 +120,19 @@ var CreditCardPaymentBox = React.createClass( {
 					cart={ this.props.cart }
 					transactionStep={ this.props.transactionStep } />
 
-				{ cartValues.isPayPalExpressEnabled( cart )
-					? <a className={ paypalButtonClasses } href="" onClick={ this.handleToggle }>
+				{ this.props.paymentMethods > 1
+					? <span>{ this.props.translate( 'Pay with:' ) }</span>
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/credit-card-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Pay with' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/checkout/checkout/ideal-payment-box.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	handleToggle = ( event ) => {
+		event.preventDefault();
+		this.props.onToggle( 'credit-card' );
+	};
+
+	render() {
+		return (
+			<PaymentBox
+				classSet="ideal-payment-box"
+				title={ translate( 'Secure Payment with iDEAL' ) }>
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/ideal-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **50** times:
`translate( 'Secure Payment with PayPal' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/payment-ideal)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/checkout/checkout/alternative-payment-methods.jsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	render() {
+		if ( ! this.props.paymentMethods ) {
+			return null;
+		}
+
+		return (
+			<div className="checkout__alternative-payment-methods">
+				<span>{ translate( 'Or pay with:' ) }</span>
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/alternative-payment-methods.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Pay with' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+				return 'iDEAL';
+		}
+
+		return this.props.paymentType;
+	}
+
+	getTitle() {
+		switch ( this.props.paymentType ) {
+			case 'ideal':
+				return translate( 'Secure Payment with iDEAL' );
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **50** times:
`translate( 'Secure Payment with PayPal' )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/payment-ideal)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+	renderButtonText() {
+		if ( cartValues.cartItems.hasRenewalItem( this.props.cart ) ) {
+			return translate( 'Purchase %(price)s subscription with %(paymentProvider)s', {
+				args: { price: this.props.cart.total_cost_display, paymentProvider: this.getPaymentProviderName() },
+				context: 'Pay button on /checkout'
+			} );
+		}
+
+		return translate( 'Pay %(price)s with %(paymentProvider)s', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: :information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+				analytics.ga.recordEvent( 'Upgrades', 'Clicked Checkout With Source Payment Button' );
+				analytics.tracks.recordEvent( 'calypso_checkout_with_source_' + this.props.paymentType );
+				location.href = sourcePaymentUrl;
+			}
+		}.bind( this ) );
+	}
+
+	renderButtonText() {
+		if ( cartValues.cartItems.hasRenewalItem( this.props.cart ) ) {
+			return translate( 'Purchase %(price)s subscription with %(paymentProvider)s', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: :information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

**Comment by akirk** [comment]
I have now successfully tested a iDEAL transaction with this PR by using the Store sandbox.

**Comment by akirk** [comment]
I have added the iDEAL logo and copied the credit card svg to get a stronger one. Here are the three states, could I get a design opinion on this? Thanks!

screen shot 2017-09-08 at 11 39 35
screen shot 2017-09-08 at 11 39 21
screen shot 2017-09-08 at 11 39 16

**Comment by yoavf** [comment]
Working great! Haven't reviewed the code yet. What do you think about launching this with an AB Test?

**Comment by yoavf** [comment]
For the iDEAL form, let's make it clarify that the name is for the receipt.
Also, the blurb below the form mentions "recurring form of payment" which is wrong for iDEAL.

**Comment by folletto** [comment]
> I have added the iDEAL logo and copied the credit card svg to get a stronger one. Here are the three states, could I get a design opinion on this? Thanks!

Heya!

This is a bit difficult, the logos don't seem clickable, so it might not be very discoverable. Would work better if we can put them inside our standard white buttons (not 100% sure... might still look odd). 

Another approach could be to put the logos in the top bar... something like this ugly mock:

screen shot 2017-09-21 at 13 35 51

If we do this it should also avoid the "pop" animation.

Another alternative: do we have to use logos? If we use plain text, with underline, would work well too.

**Comment by akirk** [comment]
Thanks @folletto that is very helpful! Logos bring quite a bit higher conversion rate, so it would be good to continue to use them. I like the tabs a lot so I'll update the PR accordingly.

**Comment by akirk** [comment]
This still needs some work on styling but it now works like this:
payment-methods
cc @yoavf

**Comment by yoavf** [comment]
I updated the design a bit and tried to make the title work on mobile (`sectionNav` being responsive).

Desktop:
bqtdjmsdpl

Mobile
azvdvbe5cq

**Comment by yoavf** [comment]
In mobile, the fact that the payment method name/logo isn't visible after the selection is problematic.

**Comment by akirk** [comment]
@folletto your UI mockup was implemented and already merged in #19006. Thank you!

**Comment by yoavf** [comment]
The CI failure is on a CSS namespace issue. I chose to ignore it to prevent adding duplication to the css, or having to go and fix a bunch of places where the existing class name is used.

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 7]
```diff
@@ -134,6 +136,28 @@ class SourcePaymentBox extends PureComponent {
 		} );
 	}
 
+	renderBankOptions() {
+		const idealBanks = {
```
Comment (akirk) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: Please add a comment where the list is from (i.e. https://stripe.com/docs/sources/ideal#optional-specifying-the-customers-bank)

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 8]
```diff
@@ -151,6 +175,17 @@ class SourcePaymentBox extends PureComponent {
 						onChange={ this.handleChange }
 						label={ translate( 'Your Name' ) }
 						eventFormName="Checkout Form" />
+					<div className="checkout-field">
+						<FormLabel htmlFor="ideal-bank">
+							{ translate( 'Bank' ) }
+						</FormLabel>
+						<FormSelect
+							name="ideal-bank"
+							onChange={ this.handleChange }
```
Comment (akirk) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: Shouldn't we have `handleChange` verify that not `-` was chosen? While I understand it should work without bank, it's still a bit pointless to allow the form to be sumitted with `Select bank...` in the dropdown selected. I understand being validated on the server-side at least for not being empty, so probably this is a _maybe_.

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 9]
```diff
@@ -134,6 +136,28 @@ class SourcePaymentBox extends PureComponent {
 		} );
 	}
 
+	renderBankOptions() {
+		const idealBanks = {
+			'-': translate( 'Select bank…' ),
```
Comment (akirk) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: What about `Please select your bank...`?

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 10]
```diff
@@ -134,6 +136,28 @@ class SourcePaymentBox extends PureComponent {
...
+			ing: 'ING',
+			knab: 'Knab',
+			rabobank: 'Rabobank',
+			regiobank: 'RegioBank',
+			sns_bank: 'SNS Bank',
+			triodos_bank: 'Triodos Bank',
+			van_lanschot: 'Van Lanschot',
+		};
+
+		return map( idealBanks, ( text, optionValue ) => (
```
Comment (akirk) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: What about we remove the `-` entry from the object and do it like this?
```
	return (
		<option value="-" key="-">
			{ translate( 'Please select your bank…' ) }
		</option>
		{ map( idealBanks, ( text, optionValue ) => (
			<option value={ optionValue } key={ optionValue }>
				{ text }
			</option>
		)}
	);
```

Comment (yoavf) [reply client/my-sites/checkout/checkout/source-payment-box.jsx]: 👍

File: client/my-sites/checkout/checkout/source-payment-box.jsx (hunk 8 above)
Comment (yoavf) [reply client/my-sites/checkout/checkout/source-payment-box.jsx]: We should add client-side validation at some point, but for now I think it's ok to rely on the server side (like we do with Paypal). W

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+		return (
+			<form onSubmit={ this.redirectToPayment }>
+
+				<div className="checkout__payment-box-section">
+					<Input
+						additionalClasses="checkout-field"
+						name="name"
+						onChange={ this.handleChange }
+						label={ translate( 'Your Name' ) }
```
Comment (akirk) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: @yoavf If you leave this empty the error message says `Invalid or missing account name` (when we actually just check for "empty"). Should we therefore rename this to `Account Name`?

**Comment by yoavf** [comment]
Getting ready for merge - rebased on `update/disable-payment-tabs-abtest` and made iDEAL the 2nd payment method rather than the default in NL

File: client/my-sites/checkout/checkout/source-payment-box.jsx [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+			triodos_bank: 'Triodos Bank',
+			van_lanschot: 'Van Lanschot',
+		};
+
+		const idealBanksOptions = map( idealBanks, ( text, optionValue ) => (
+			<option value={ optionValue } key={ optionValue }>{ text }</option>
+		) );
+
+		return [
+			<option value="" key="-">{ translate( 'Please select your bank…' ) }</option>,
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/source-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **1** time:
`translate( 'Please select your bank.' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Pull request merged**